File: src/utils/file.ts
import { promises as fs } from "fs";
import path from "path";
import { paths } from "../config/paths";
import { CodeChange, FileSnapshot } from "../core/types";

export async function writeChanges(changes: CodeChange[]): Promise<void> {
  for (const change of normalizeRepoChanges(changes)) {
    const fullPath = path.join(paths.rootDir, change.path);
    await fs.mkdir(path.dirname(fullPath), { recursive: true });
    await fs.writeFile(fullPath, change.content, "utf8");
  }
}

export async function applyChangesWithSnapshots(
  changes: CodeChange[],
): Promise<FileSnapshot[]> {
  const snapshots: FileSnapshot[] = [];

  for (const change of normalizeRepoChanges(changes)) {
    const fullPath = path.join(paths.rootDir, change.path);
    const previousContent = await readFileIfExists(fullPath);
    await fs.mkdir(path.dirname(fullPath), { recursive: true });
    await fs.writeFile(fullPath, change.content, "utf8");

    snapshots.push({
      path: change.path,
      previousContent,
      nextContent: change.content,
    });
  }

  return snapshots;
}

export async function restoreSnapshots(
  snapshots: FileSnapshot[],
): Promise<void> {
  for (const snapshot of snapshots) {
    const fullPath = path.join(paths.rootDir, snapshot.path);

    if (snapshot.previousContent === null) {
      await fs.rm(fullPath, { force: true });
      continue;
    }

    await fs.mkdir(path.dirname(fullPath), { recursive: true });
    await fs.writeFile(fullPath, snapshot.previousContent, "utf8");
  }
}

export async function readWorkspaceFile(relativePath: string): Promise<string> {
  const safePath = normalizeRepoPath(relativePath);
  const fullPath = path.join(paths.rootDir, safePath);
  return fs.readFile(fullPath, "utf8");
}

async function readFileIfExists(fullPath: string): Promise<string | null> {
  try {
    return await fs.readFile(fullPath, "utf8");
  } catch {
    return null;
  }
}

function normalizeRepoChanges(changes: CodeChange[]): CodeChange[] {
  return changes.map((change) => ({
    ...change,
    path: normalizeRepoPath(change.path),
  }));
}

function normalizeRepoPath(inputPath: string): string {
  const unixPath = inputPath.replace(/\\/g, "/").replace(/^\/+/, "");
  const withoutRepoPrefix = unixPath
    .replace(/^repo\/app\//, "")
    .replace(/^repo\//, "")
    .replace(/^app\//, "");
  const candidate = path.posix.normalize(`repo/app/${withoutRepoPrefix}`);

  if (!candidate.startsWith("repo/app/")) {
    throw new Error(`Refusing to write outside repo/app: ${inputPath}`);
  }

  return candidate;
}

File: src/utils/file.py
"""File utilities — replaces file.ts."""
from __future__ import annotations
from pathlib import Path
from ..config.paths import paths
from ..core.types import CodeChange, FileSnapshot

def _normalize_repo_path(input_path: str) -> str:
    unix = input_path.replace("\\", "/").lstrip("/")
    clean = unix.replace("repo/app/", "").replace("repo/", "").replace("app/", "")
    candidate = f"repo/app/{clean}"
    if not candidate.startswith("repo/app/"): raise ValueError(f"Refusing to write outside repo/app: {input_path}")
    return candidate

async def write_changes(changes: list[CodeChange]) -> None:
    for c in changes:
        p = paths["root_dir"] / _normalize_repo_path(c.path)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(c.content, encoding="utf-8")

async def apply_changes_with_snapshots(changes: list[CodeChange]) -> list[FileSnapshot]:
    snaps = []
    for c in changes:
        p = paths["root_dir"] / _normalize_repo_path(c.path)
        prev = p.read_text(encoding="utf-8") if p.exists() else None
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(c.content, encoding="utf-8")
        snaps.append(FileSnapshot(c.path, prev, c.content))
    return snaps

async def restore_snapshots(snapshots: list[FileSnapshot]) -> None:
    for s in snapshots:
        p = paths["root_dir"] / _normalize_repo_path(s.path)
        if s.previous_content is None:
            if p.exists(): p.unlink()
        else:
            p.parent.mkdir(parents=True, exist_ok=True)
            p.write_text(s.previous_content, encoding="utf-8")

async def read_workspace_file(rel_path: str) -> str:
    return (paths["root_dir"] / _normalize_repo_path(rel_path)).read_text(encoding="utf-8")
